### ISSUE TITLE
Add mzworkflows.py to support arbitrary Python workflows instead of being restricted to YAML

### ISSUE THREAD

**Issue by cirego** [MaterializeInc/materialize#6439]
> we really want to be writing programs for workflows.

100% agree. Certainly for the workloads of any notable complexity.

>  I hadn't thought about the possibility of having mzcompose take a script as a workflow, either explicitly or implicitly, as an argument.

Yeah, I was thinking that'd be the easiest way to mash up the YAML and Python worlds. I actually remember now that I started mocking up this idea like so:

```py
# mzworkflows.py

def workflow_one(composition: mzcompose.Composition, args: Iterable[str]):
    pass

def workflow_two(composition: mzcompose.Composition, args: Iterable[str]):
    pass
```

The idea being you could drop an `mzworkflows.py` file into any directory with an mzcompose.yml file. Then mzcompose would use `importlib.import_module` to attempt to load that file. Any functions in that file whose name began with `workflow_` and took the right kind of parameters would be considered to be workflows, just like workflows defined in YAML. You'd be able to call them directly with `./mzcompose run`, or use them from a YAML WorkflowWorkflowStep, etc.

_Originally posted by @benesch in https://github.com/MaterializeInc/materialize/issues/5588#issuecomment-775679619_

**Comment by benesch** [comment]
Sweeet, let's do it! I think the `importlib` bit is probably the easy part. The hard part was converting all the existing `WorkflowStep` classes into methods on `composition` that can be called easily from Python.

**Comment by philip-stoev** [comment]
I am going to take a stab at this with a design document.

**Comment by philip-stoev** [comment]
Closed #6439 via #7797 and #7841
